<?xml version="1.0" ?>
<launch>
  <arg name="robot_ip"     default="192.168.3.102"/>
  <arg name="arm_id"       default="panda" />
  <arg name="load_gripper" default="true" />

  <param name="robot_description"
       command="xacro $(find franka_description)/robots/panda/panda.urdf.xacro hand:=$(arg load_gripper) arm_id:=$(arg arm_id)" />

  <include file="$(find franka_gripper)/launch/franka_gripper.launch" if="$(arg load_gripper)">
    <arg name="robot_ip" value="$(arg robot_ip)" />
    <arg name="arm_id"   value="$(arg arm_id)" />
  </include>

  <node name="franka_control" pkg="franka_control" type="franka_control_node" output="screen" required="true">
    <rosparam command="load" file="$(find franka_interactive_controllers)/config/franka_control_node_interactive.yaml" subst_value="true" />
    <param name="robot_ip" value="$(arg robot_ip)" />
  </node>

  <rosparam command="load" file="$(find franka_interactive_controllers)/config/default_controllers_interactive.yaml" subst_value="true" />

  <node name="state_controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="franka_state_controller"/>
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen"/>
  <node name="joint_state_publisher" type="joint_state_publisher" pkg="joint_state_publisher" output="screen">
    <rosparam if="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states, franka_gripper/joint_states] </rosparam>
    <rosparam unless="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states] </rosparam>
    <param name="rate" value="1000"/>
  </node>
  <node name="joint_state_desired_publisher" type="joint_state_publisher" pkg="joint_state_publisher" output="screen">
    <rosparam if="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states_desired, franka_gripper/joint_states] </rosparam>
    <rosparam unless="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states_desired] </rosparam>
    <param name="rate" value="1000"/>
    <remap from="/joint_states" to="/joint_states_desired" />
  </node>


  <!-- Loads controller parameters -->  
  <rosparam command="load" file="$(find franka_interactive_controllers)/config/impedance_control_additional_params.yaml"/> 
  <rosparam command="load" file="$(find franka_interactive_controllers)/config/franka_interactive_controllers.yaml" />
  
</launch>
